### ISSUE TITLE
How do you make the median object size within the field of view of the NN?

### ISSUE THREAD

**Issue by mballari** [mpicbg-csbd/stardist#53]
Hi,

When I run the 2D training notebook using my own data, I get the following warning:
"WARNING: median object size larger than field of view of the neural network."

What does this mean exactly, and how can I make it so that median object size is within the field of view of the neural network? Is it an issue with the data I am trying to train the model with? When I run the training notebook all the way, the IoU is 0, and I'm thinking that the reason is because of the median object size being larger than the field of view of the neural network. 

Sincerely,
Meg

**Comment by maweigert** [comment]
> What does this mean exactly, and how can I make it so that median object size is within the field of view of the neural network?

The field of view (or receptive field) of a neural network is the region of the input that affects the value of a single output pixel. In general the deeper the network and the bigger the convolutional kernels inside each layer, the larger the receptive field. For the 2D StarDist model from the [example notebooks](https://nbviewer.jupyter.org/github/mpicbg-csbd/stardist/tree/master/examples/2D) the receptive field is roughly 180x180. 
What you can do in your example is either downscaling your data (such that the diameter of your annotated cell segmentations becomes smaller than ~100 pixels) or increase the `grid` parameter in the config, e.g setting `grid=(4,4)`.    
You could as well post some example image of your data such that it is easier to understand the issue.

All the best,
Martin

**Comment by mballari** [comment]
Thank you, Martin! Increasing the grid parameter to (4, 4) made the median object size within the field of view of the neural network, but the mean true score is still 0 as shown below. I have also attached an example of an image and its corresponding label from my data.

![image](https://user-images.githubusercontent.com/59752040/80929079-02b3c700-8d5e-11ea-8c87-54382ee070c1.png)

![image](https://user-images.githubusercontent.com/59752040/80929047-d7c97300-8d5d-11ea-8f14-7870b6b1733e.png)

**Comment by maweigert** [comment]
Could you share the full output of your trainings notebook (e.g. download as html file)?

**Comment by mballari** [comment]
Attached is a pdf of my training notebook!

[Meg_StarDist_Training.pdf](https://github.com/mpicbg-csbd/stardist/files/4573947/Meg_StarDist_Training.pdf)

**Comment by maweigert** [comment]
ok, thanks. As you can see from the training output:
<img width="738" alt="Screenshot 2020-05-04 at 12 16 43" src="https://user-images.githubusercontent.com/11042162/80956271-2b40cd00-8e01-11ea-9a2b-9b0413129fc8.png">

the validation loss is extremely large already after the first epoch, which indicates that something is wrong with your validation images. That's why looking at the loss curves in tensorboard is the first thing to do when something goes wrong: Is both the loss and the validation loss decreasing?  

Some things to ponder:
- Do the validation images contain labeled masks at all?   
- What is the size of your raw images? Maybe increasing the patch_size in the Config object might help: 

```python

conf = Config2D (...
    train_patch_size = (720, 720),
...)
```

**Comment by mballari** [comment]
My validation set contained 8 images, 3 of which had labels that were heavily annotated with nuclei, while the other 5 labels had 0 nuclei. I tried manually choosing the validation images so that 6 of the images would have labels annotated with nuclei, while 2 had 0 nuclei,  but the validation loss was still extremely large. Also my images are 256 x 256, so I couldn't increase the patch_size.

**Comment by maweigert** [comment]
Check, whether the validation stacks `X_val, Y_val` are somehow look different than the training data (e.g. normalisation). How did you generate the validation data, via the notebook? There seems to be clearly something wrong with them...

**Comment by mballari** [comment]
Hi Martin, I looked over the validation images that were generated via the notebook like you asked and they look fine, and have been normalized similar to the training images. I can't seem to figure out what is wrong with them. What I am confused about though is that the array for the validation images ind_val = [53 20 7 42 14 28 51 38], but then when I had the figures of the images shown, they corresponded to images I had titled in my folder as [9 27 15 47 21 34 7 43]. But all the labels still matched the raw images.

**Comment by maweigert** [comment]
You could try the following 

```python 
conf = Config2D(
...
    train_foreground_only= 1,
    train_background_reg= 0 
...
```

But in general, your images seem to be pretty hard:

![image](https://user-images.githubusercontent.com/11042162/81592891-eb5c8580-93be-11ea-9c47-f1e51eab93fc.png)

So I wouldn't be too optimistic...

**Comment by mballari** [comment]
Do you have any suggestions for making my raw images easier to read?

**Comment by maweigert** [comment]
> Do you have any suggestions for making my raw images easier to read?

I would try to see whether you can improve your imaging pipeline - currently the contrast seems to be so low that is even hard to reliably detect cell outlines by eye...

**Comment by mballari** [comment]
Ok, thank you so much for all your help, Martin!!

**Comment by DaniK-BIP** [comment]
Hello,
Thank you all for the great work!

I have the same problem, however reversed. My images and masks have the dimensions of 54x61 but i get the same warning "WARNING: median object size larger than field of view of the neural network."
Would anyone know an answer to this?

**Comment by maweigert** [comment]
HI @DaniK-BIP,

hard to tell without more information. Can you post an example image/annotation pair and your model config?
